model/src/main/java/uk/gov/pay/commons/api/validation/ExternalJsonMetadataValidator.java

```diff
@@ -1,93 +1,90 @@
+        return errors;
+    };
+
+    private static final BiFunction<JsonNode, List<String>, List<String>> VALIDATE_MAX_NUMBER_OF_KEY_PAIRS = (jsonNode, errors) -> {
+        if (jsonNode.size() > MAXIMUM_KEYS)
+            return arrayList(format("cannot have more than %d key-value pairs", MAXIMUM_KEYS));
+        return errors;
+    };
+
+    private static final BiFunction<JsonNode, List<String>, List<String>> VALIDATE_KEYS_ARE_NOT_NULL = (jsonNode, errors) -> {
+        var blankOrNullKeys = Stream.iteratorStream(jsonNode.fields())
+                .filter(entry -> entry.getKey() == null)
+                .toList();
+        if (blankOrNullKeys.length() > 0) return arrayList("keys must not be null").append(errors);
+        return errors;
+    };
+
+    private static final BiFunction<JsonNode, List<String>, List<String>> VALIDATE_KEY_LENGTHS = (jsonNode, errors) ->
+            Stream.iteratorStream(jsonNode.fields())
+                    .filter(entry -> entry.getKey() != null && keyLengthIsInvalid(entry.getKey()))
+                    .foldLeft((accumulatedErrors, entry) ->
+                                    arrayList(format("keys must be between 1 and %d characters long", MAXIMUM_KEY_LENGTH)).append(accumulatedErrors),
+                            errors);
+
+    private static final BiFunction<JsonNode, List<String>, List<String>> VALIDATE_VALUE_TYPES = (jsonNode, errors) ->
+            Stream.iteratorStream(jsonNode.fields())
+                    .filter(entry -> valueTypeIsNotAllowed(entry.getValue()))
+                    .foldLeft((accumulatedErrors, entry) ->
+                                    arrayList(format("value for '%s' must be of type string, boolean or number", entry.getKey())).append(accumulatedErrors),
+                            errors);
+
+    private static final BiFunction<JsonNode, List<String>, List<String>> validateValueLengths = (jsonNode, errors) ->
+            Stream.iteratorStream(jsonNode.fields())
+                    .filter(entry -> entry.getValue().asText().length() > MAXIMUM_VALUE_LENGTH)
+                    .foldLeft((accumulatedErrors, entry) ->
+                                    arrayList(format("value for '%s' must be a maximum of %d characters", entry.getKey(), MAXIMUM_VALUE_LENGTH)).append(accumulatedErrors),
+                            errors);
+
     @Override
     public boolean isValid(JsonNode jsonNode, ConstraintValidatorContext constraintValidatorContext) {
 
-        if (jsonNode == null) {
-            return true;
-        }
+        if (jsonNode == null) return true;
 
         constraintValidatorContext.disableDefaultConstraintViolation();
 
-        if (jsonNode.isArray()) {
-            addConstraint("must be an object of JSON key-value pairs", constraintValidatorContext);
-            return false;
-        }
-
-        var errorMessages = validateKeyValuePairs(jsonNode);
+        List<String> errors =
```

Comment:
This could be a `Set<String>` since we only report each distinct error once. The validate method will return a `Set<ConstraintViolation<T>>` and each violation will be equal if the message is the same. Doesn't hurt as a List but just means it might try to create a violation that will ultimately be ignored if its already there.